### PR TITLE
CV terms on all fields

### PR DESCRIPTION
**Pull request by spficklin**
# Tripal 4 Core Dev Task

## Description

Currently, users can only set CV terms on Tripal fields. But someone can attach an existing Drupal field to a Tripal content type. For example, if a user wants to add an image to an organism content type they can do so using the Drupal image field, but, that image field doesn't have a CV term associated with it. For web services to work correctly all fields attached to a Tripal content type must be associated with a CV term.

This PR fixes the problem by using a `hook_form_alter()` function to add the CV term form elements to any field settings page that is attached to a tripal entity type.    It also sets a 3rd party setting with the field so that later, when we setup web services, we can get the CV term from the field.  

## Testing?

1. On an existing Tripal site, go to admin > Structure > Tripal Content Types and select "manage fields" for the "Organism" content type.
2. Click the 'Create a New Field' button and in the page that appears, add an Image field (Drupal 10.2 first select "File upload"). Name it "Organism Image".
3. When editing the image you will first see the field settings page.  Click the "save field settings button"
4. You will recieve the warning:  "The field is missing an assigned controlled vocabulary term. Please set one"
5. Scroll to the section "Controlled Vocabulary Section" and add the term "Image (data:2968)"
6. Click the "save settings" button at the bottom.  You now have a new image field. 
7. Click the link to edit the new "Organism Image" field. You will see that in the "Controlled Vocabulary" fieldset that the term is set and present.  You can change the term, save, and re-open and the term change is maintained.

### CODE REVIEW

File: tripal/src/TripalField/TripalFieldItemBase.php [hunk 1]
```diff
@@ -149,24 +149,40 @@ protected function buildVocabularyTermTable(array &$elements,
...
     $is_open = FALSE;
     $term = NULL;
     $idSpace = NULL;
     $vocabulary = NULL;
-    $termIdSpace = $this->getSetting('termIdSpace');
-    $termAccession = $this->getSetting('termAccession');
-    $debug = $this->getSetting('debug');
+    $debug = $field->getSetting('debug');
+
+    // IF this field is not a TripalField then we wnat to add a small
```
Comment (laceysanderson) [inline tripal/src/TripalField/TripalFieldItemBase.php]:
```suggestion
    // IF this field is not a TripalField then we want to add a small
```

**Review by laceysanderson — COMMENTED** [review]
So apparently Drupal 10.2.x has brought some pretty big changes to the UI for adding custom fields to an entity and they impact this PR.

Whereas before we had separate pages for field settings and storage settings, in Drupal 10.2.x they have all been combined into a single page.

✅ Term form is rendered on the image field settings page
✅ Term is saved and is parsed into the table format on edit
❌ There is an error shown when the field is created/edited

createField
singlePageCreateFieldForm
errorOnSubmit
organismFieldEdit

**Comment by spficklin** [comment]
> Whereas before we had separate pages for field settings and storage settings, in Drupal 10.2.x they have all been combined into a single page.

:scream:

That emoji needs to be bigger.

**Comment by dsenalik** [comment]
image

**Comment by dsenalik** [comment]
~~Ignoring the scream part, I tested on drupal 10.1 and at step 4 of testing I am not seeing the described new cvterm related things...~~
~~checking, and `tripal_form_field_config_edit_form_alter()` is never being called~~
I started up my docker incorrectly, I appear to be testing the "before" state 😐
With the correct command, things are working up to step 6 where I see
`Warning: Undefined array key "#is_tripal_field" in Drupal\tripal\TripalField\TripalFieldItemBase::fieldSettingsFormValidate() (line 300 of modules/contrib/tripal/tripal/src/TripalField/TripalFieldItemBase.php). `

**Comment by dsenalik** [comment]
Here is the reference for the "scream" changes
https://www.drupal.org/node/3386675
**Field configuration forms restructured: Field Storage Configuration Form is now a subform within Field Configuration Edit Form**
Introduced in branch: 10.2.x
Introduced in version: 10.2.0
Description: 
The \Drupal\field_ui\Form\FieldStorageConfigEditForm is now rendered as a subform within \Drupal\field_ui\Form\FieldConfigEditForm. This structural change allows for a more streamlined configuration experience. However, this restructuring necessitates updates in form alters and field configuration updates.
...

**Comment by laceysanderson** [comment]
Note: This has also broken the form for all Tripal and Chado fields 🤦‍♀️ You can see progress on this in issue #1791 and branch `tv4g1-1791-editFieldWSOD`.

File: tripal/src/TripalField/TripalFieldItemBase.php [hunk 2]
```diff
@@ -149,24 +149,40 @@ protected function buildVocabularyTermTable(array &$elements,
...
-    $termIdSpace = $this->getSetting('termIdSpace');
-    $termAccession = $this->getSetting('termAccession');
-    $debug = $this->getSetting('debug');
+    $debug = $field->getSetting('debug');
+
+    // IF this field is not a TripalField then we want to add a small
+    // tag to the field so that we know in the validate we need to add
+    // a setting for the cv term for this field.
+    if (!is_subclass_of($field, 'Drupal\tripal\TripalField\TripalFieldItemBase')) {
+      $form['#is_tripal_field'] = FALSE;
```
Comment (dsenalik) [inline tripal/src/TripalField/TripalFieldItemBase.php]:
```suggestion
      $elements['is_tripal_field'] = [
        '#type' => 'hidden',
        '#value' => 0,
      ];

```

File: tripal/src/TripalField/TripalFieldItemBase.php [hunk 3]
```diff
@@ -149,24 +149,40 @@ protected function buildVocabularyTermTable(array &$elements,
...
+      $termIdSpace = $field->getThirdPartySetting('tripal', 'termIdSpace');
+      $termAccession = $field->getThirdPartySetting('tripal', 'termAccession');
+    }
+    else {
+      $form['#is_tripal_field'] = TRUE;
```
Comment (dsenalik) [inline tripal/src/TripalField/TripalFieldItemBase.php]:
```suggestion
      $elements['is_tripal_field'] = [
        '#type' => 'hidden',
        '#value' => 1,
      ];
```

File: tripal/src/TripalField/TripalFieldItemBase.php [hunk 4]
```diff
@@ -270,9 +294,17 @@ public static function fieldSettingsFormValidate(array $form, FormStateInterface
       $accession = $matches[3];
       $form_state->setValue(['settings','termIdSpace'], $idSpace_name);
       $form_state->setValue(['settings','termAccession'], $accession);
+
+      // If this isn't a Tripal field then add a third party setting so
+      // we know what the cvterm is.
+      if ($form['#is_tripal_field'] == FALSE) {
```
Comment (dsenalik) [inline tripal/src/TripalField/TripalFieldItemBase.php]:
```suggestion
      if ($settings['is_tripal_field'] == 0) {
```

**Review by dsenalik — DISMISSED** [review]
These are the changes I made to get this to work for me, since changes to `$form` are not passed back from the function. See what you think @spficklin

**Comment by laceysanderson** [comment]
This is on hold until we get things working with the field settings subform in D10.2

**Comment by dsenalik** [comment]
I merged in 4.x and fixed conflicts.
For my testing I included my Feb. 20 review suggestions.
I tested on Drupal 10.2 and it wasn't working in that it did not save the term.
I saw what was wrong so I merged in my Feb. 20 suggestions https://github.com/tripal/tripal/pull/1790/commits/cd0b1e01c3aabc97afc507213b500f4172611168, and then fixed it in commit https://github.com/tripal/tripal/pull/1790/commits/59f962bb16020b156739ac2654da6085d2c82b8f 

Tested and working on both Drupal 10.2 and 10.1

Of course now I am no longer "impartial" so maybe @Lacey will be willing to review this? Not nearly as hard as that last one! 😁

**Review by laceysanderson — APPROVED** [review]
Works beautifully now!

I tested on Drupal 10.1 and 10.2. Specifically, I 
1. added a Drupal image field to the organism content type + assigned cvterm
2. edited the field and confirmed cvterm was still there
3. created an organism and confirmed the image field works
4. added an image field to a Drupal Article and confirmed it did not ask for a cvterm
5. edited and confirmed it still did not ask for  cvterm
6. created an article and confirmed that it worked
7. checked a random sub sampling of chado fields and confirmed their forms still looked right and saved without error out of paranoia 😉  there was still a cvterm on these forms and it was unaffected as expected 🤪

**Comment by dsenalik** [comment]
Thanks for the extra thorough testing! 😃

**Pull request merged**